torchgeo/datamodules/spacenet.py

```diff
@@ -0,0 +1,183 @@
+
+# https://github.com/pytorch/pytorch/issues/60979
+# https://github.com/pytorch/pytorch/pull/61045
+DataLoader.__module__ = "torch.utils.data"
+
+
+class SpaceNet1DataModule(pl.LightningDataModule):
+    """LightningDataModule implementation for the SpaceNet1 dataset.
+
+    Randomly splits into train/val/test.
+
+    .. versionadded:: 0.4
+    """
+
+    def __init__(
+        self,
+        batch_size: int = 64,
+        num_workers: int = 0,
+        val_split_pct: float = 0.1,
+        test_split_pct: float = 0.2,
+        **kwargs: Any,
+    ) -> None:
+        """Initialize a LightningDataModule for SpaceNet1.
+
+        Args:
+            batch_size: The batch size to use in all created DataLoaders
+            num_workers: The number of workers to use in all created DataLoaders
+            val_split_pct: What percentage of the dataset to use as a validation set
+            test_split_pct: What percentage of the dataset to use as a test set
+            **kwargs: Additional keyword arguments passed to
+                :class:`~torchgeo.datasets.SpaceNet1`
+        """
+        super().__init__()
+        self.batch_size = batch_size
+        self.num_workers = num_workers
+        self.val_split_pct = val_split_pct
+        self.test_split_pct = test_split_pct
+        self.kwargs = kwargs
+
+        self.padto = K.PadTo((448, 448))
+
+    def on_after_batch_transfer(
+        self, batch: Dict[str, Any], batch_idx: int
+    ) -> Dict[str, Any]:
+        """Apply batch augmentations after batch is transferred to the device.
+
+        Args:
+            batch: mini-batch of data
+            batch_idx: batch index
+
+        Returns:
+            augmented mini-batch
+        """
+        if (
+            hasattr(self, "trainer")
+            and self.trainer is not None
+            and hasattr(self.trainer, "training")
+            and self.trainer.training
+        ):
+            # Kornia expects masks to be floats with a channel dimension
+            x = batch["image"]
+            y = batch["mask"].float().unsqueeze(1)
+
+            train_augmentations = K.AugmentationSequential(
+                K.RandomRotation(p=0.5, degrees=90),
+                K.RandomHorizontalFlip(p=0.5),
+                K.RandomVerticalFlip(p=0.5),
+                K.RandomSharpness(p=0.5),
+                K.ColorJitter(
+                    p=0.5,
+                    brightness=0.1,
+                    contrast=0.1,
+                    saturation=0.1,
+                    hue=0.1,
+                    silence_instantiation_warning=True,
+                ),
+                data_keys=["input", "mask"],
+            )
+            x, y = train_augmentations(x, y)
+
+            # torchmetrics expects masks to be longs without a channel dimension
+            batch["image"] = x
+            batch["mask"] = y.squeeze(1).long()
+
+        return batch
+
+    def preprocess(self, sample: Dict[str, Any]) -> Dict[str, Any]:
```

Comment:
I clearly haven't written a datamodule in a while. Is there a reason to do this here instead of in `on_after_batch_transfer`?

Reply:
No worries!

`on_after_batch_transfer` only happens when transferring to the GPU for the training set while `preprocess` happens to every sample

Reply:
Only the training set is run on the GPU? Or on_after_batch_transfer is only run for the training set?

Reply:
`on_after_batch_transfer` is run when any batch is transferred from the CPU to the GPU (see https://pytorch-lightning.readthedocs.io/en/latest/data/datamodule.html#on-after-batch-transfer), we currently have checks in place to only do the augmentation when we are training (not val, test, or inference). This is cool because it lets us do augmentation logic directly on the GPU.

I like to separate the image preprocessing to make it clear what you need to do to the images to use the models in inference mode. I.e. if we release trained models that have used this SpaceNet1DataModule, it is quite clear what inputs the model expects (SpaceNet1 dataset output fed through `preprocess`)

Reply:
Personally I would run everything on the GPU so that `on_after_batch_transfer` runs normalization and optionally runs augmentation when training. But this is the same as all of our existing datamodules, so we can consider changing this at a later date.